### ISSUE TITLE
Implement standardized/modularized question types

### ISSUE THREAD

**Issue by dkoslicki** [RTXteam/RTX#52]
See #39 for discussion. The main idea is to make individual classes (or some other data structure) for the different question types. Should include methods such as:
``restate_question`` (return human readable restatement of question, with arguments filled in slots) 
``give_example`` (return a couple examples of what questions this can answer)
``extract_parameters`` (extract the actual arguments from an input question)
``format_answer`` (format the response for parsing by RTXQuery)

and include properties such as:
``type_list`` specifying the required bioentity types (including relationships) and any conditions on them (such as requiring a certain disease class as input).
``argument_list`` specifying the extracted parameters (conforming to ``type_list``) to be based to the question answering script.
``question_corpus`` a set of example phrasings of the question for the NLP to work.

**Comment by dkoslicki** [comment]
Will need to figure out naming convention for ease of translating between multiple query types and the solution scripts. For example: "what proteins does naproxen target" and "what are the phenotypes of malaria" will both point to the same script. What should this script be called?

**Comment by edeutsch** [comment]
How about:
QMASolution.py
for query multiple A
QMBSolution.py
QMCSolution.py
...

**Comment by dkoslicki** [comment]
That may work, but would there be a clear mapping between:
Q5, Q6, Q7 - > QMA
Q8, Q9 -> QMB

Maybe we should adopt the following convention:
Questions have short, common language identifiers, appended with an integer to specify multiple copies. So for example:
Q1 becomes QDiseaseProtects
Q2 becomes QCOP
and the various Q3's become:
QOneHop1, QOneHop2, ...

and the similarity questions ("what diseases have similar symptoms to X?" become:
QJaccardSimilarity1, QJaccardSimilarity2,...

I imagine this will be easier to keep straight in the future (since we are soon going to deploy a bunch of new question types).

**Comment by dkoslicki** [comment]
@edeutsch Ok Eric, I've implemented a class that does what we talked about.
See [the Question class](https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/Question.py#L169-L280).
I've also [included a script](https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/ParseQuestions.py) showing how you can interact with it.

I've also added [a TSV file](https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/Questions.tsv) where we can put in all the question definitions.

Before I get too much further along, I wanted to check with you that this is on track for what we had in mind. The only complicating factor is that Q0 is parsed differently than the other question types. However, one benefit of adopting your "one question type for each node/relationship type" is that it actually makes the NLP parsing much easier (since I only need to look for actual node names, not relationship types or relationships).

A couple things to note:
1. The "restated question" column in [Questions.tsv](https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/Questions.tsv) use string templates such as "Which genetic conditions might offer protection against $disont_disease?" where the variable after $ must match a node type in the KG. The "types" column must also match a node type in the KG.
2. I added a column "Solution script" where we can keep track of which script solves which question. Cool thing is that we can probably use this to automatically call the corresponding script we need to answer the question.

The main tasks now are:
1. Make sure this is on track to be further developed
2. Create a new class that handles the conversion into the format you want (with error messages, time stamps, etc.). I figure that you would want to do this since that's the main integration point with RTXQuery.py
3. Make sure all questions are implemented into the TSV file
4. Lots and lots of testing

**Comment by edeutsch** [comment]
Yes, this looks great, this is what I had in mind. Q0 isn't actually script, but rather a method call. But maybe we should change that.

**Comment by edeutsch** [comment]
Also, I agree with your QDiseaseProtects, etc. proposal. That is a bit more sensible than my proposal.

**Comment by dkoslicki** [comment]
@edeutsch Ok, looks like we're good to go with the question templating stuff. A few things to note:
1. You can see what the output format looks like now with ``python3 ParseQuestion.py``.
2. The fields of the parsed results have now changed slightly. In particular, I have added a new one called ``executionString`` that gives the exact command required to fulfill an input query.
3. The field ``knownQueryTypeID`` has now been removed in favor of ``solutionScript`` since as we discussed above, we will no longer be referring to questions in terms of ``Q<number>``
4. The ``terms`` field is still present, but I really think we should depreciate this given the addition of ``executionString`` as we don't need to coordinate the order of the ``terms`` list.
5. If you take a look at [Questions.tsv](https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/Questions.tsv) you'll see that I updated the format to include things like: how exactly the script is called, additional parameters that are required, etc.
6. I put a placeholder ``Q0Solution.py`` to stand in for your solution to the "what is X" question. Let me know if you want to create a real ``Q0Solution.py``, or if you want me to treat this question as a special case (and so don't populate the ``executionString`` field, etc.)

Oh yeah, and now since we have a total of 31 questions, I [made a new meta-question](https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/WhatQuestionsCanYouAnswer.py) that will randomly sample from our list of questions, fill in the proper terms with random nodes in the graph and then return it to the user.

I think the next step is to make sure this plays well on your end. I think this should greatly simplify your ``RTXQuery.py`` since you won't need a bunch of if statements. You'll just need to make sure the ``executionString`` is populated (via seeing if ``errorMessage`` is None), and then call the execution string.

Let me know when it is playing well with your UI and we can close this issue.

**Comment by edeutsch** [comment]
Hi David, thanks, this looks like good progress, but I have a few concerns. few comments on your points:
1. Nice, seems mostly good, except for a few of the points below

2. hmm. I like the idea of encoding in a file the command that will be executed, but I don't think we should be sending it an receiving it in the web service. For two reasons. First, we can't pass a command to be run to query and have it just run it. That is a security nightmare waiting to happen even with attempts at safeguards. What if the user passes to the query service 'executionString': "Q0Solution.py -i 'iphone'; /bin/rm -rf /mnt/data/orangeboard". Would query just run that? And second, I think we (NCATS) are trying to grope our way toward some level of standardization, where in principle other groups could/would use our API to implement their own functionality. or we have a standard API. And this means that we shouldn't be passing data that is highly specific to our implementation. I think it is far better to pass an identifier for a semantic query type and then let the local implementation figure out how to answer it (by pulling that execution string out of the data file).

3. After pondering this a little more, I think getting rid of the Q0 notation or some sort of alternate identifier is not a good thing. First, I think it is helpful for us to be able to enumerate all the semantic query types that we can answer. Giving them identifiers will make it much easier to refer to them. We can easily talk about Q2 and Q5, etc. Using 'solutionScript': 'SimilarityQuestionSolution.py' is way too implementation specific I think. and lots of semantically different questions now have the same handle. I really think we should have an clear, implementation neutral identifier for each semantic question (again, which script is called to answer it is an implementation detail that should be hidden in the API). I still like Q1, Q2, etc. These are neutral and short, easy to say, easy to refer to. We should publish these publicly and encourage others to use it. Even if they choose not to, it would be relatively straightforward to create a mapping between Team Gamma query type identifiers and Team X-ray query type identifiers. But only if we have a clear identifier per semantic query type. I think this is important. Both QuestionParser and the QueryHandler can both refer to the same file to share implementation details, but the public API communication should be neutral.

4. I rather think we do. First, because of point 2 above, we shouldn't be passing execution strings. We could move to a hash instead of the list perhaps, but then we'll need to consider potentially having disease_1 and disease_2 as keys in case we have query types with 2 order-dependent diseases ("how does diabetes cause eye disease?")

5. This looks great. I just think we need some semi-public identifier for each of these. I'm happy with Q1 Q2, etc., but if we want something else, that's okay. I think we should aim for reusability by other groups, even if they don't end of actually doing it. We should try.

6. Sure, this is fine. We might as well just move this to an external script for consistency.

Sorry to be a bit negative here, but I feel a bit strongly on some of these issues. We're basically going in the right direction, but I think some of the details would be missteps and we should correct.

**Comment by dkoslicki** [comment]
@edeutsch You're points are well-taken, and implementing your suggestions should be relatively straight-forward. To address each:
2.+4. Ah, of course. I had in mind only local interactions with the executionString, and completely spaced about the fact that a user could technically inject something malicious in here. If you give me a format (or we settle on one) for how the parameters and solution script should be passed to the RTXQuery, I'm happy to implement it. I think the solution to multiple disease terms is as simple as you indicated: disont_disease_@1 and disont_disease_@2 or some other set of characters appended to the end of the parameter string. I can have the code split on this string (specifically '@') and the first part will be a node label and the second part will be which parameter it is.

And again, I think a hash-based/dict structure is the way to go. The order in which flags are used to call the solution script is arbitrary and hence the order in the "terms" list is arbitrary, hence difficult to standardize on. Using dicts appears to be the common way to pass parameters in python (even the string templates use: ``template.substitute(<dict>)`` to pass parameters), and there are handy constructs like ``f(**<dict>)`` to unpack terms.

3.+5. Yeah, I'm fine with numerical identifiers, especially for the reasons you indicated. It should be easy now that we are divorcing the QX terminology from the name of the script which answers it. What about just using the order given by the ``Questions.tsv`` file? I could add explicit row numbers in there to make it even clearer...

**Comment by edeutsch** [comment]
Great, thanks. So for implementation changes from your current state, I would:
1) Add as the first column in your data file the queryTypeId strings, whatever they are. I think preserving Q1 and Q2 as an acknowledgement of their history and place in all our minds is useful, but not necessary. We could alternatively call them out as FOAQ1 and FOAQ2 and give the rest just Q3, Q4, etc. I am fine to renumber current Q0 if desired. I thought it was cute, but others found starting at 0 abhorrent. I would suggest we preserve what we originally had, but am okay with slight changes to suit.

2) Here's how I would adjust the QuestionTranslator script response:
a) Keep originalQuestion and restatedQuestion as currently done. Or, I suppose, since we're standardizing on snake_case, let's make that original_question and restated_question. 
b) Remove solutionScript. Remove executionString. (this is from the response, not from the file!)
c) Restore knownQueryTypeID. Or, for snake_case known_query_type_id
d) We can change terms to a dict, okay. We do need to be careful about the types here. We don't really want disont_disease, we want just disease. But if the KG hasn't changed yet, then we can hang on a little more. But the plan is to have the file and all the code use disease instead of disont_disease and chemical_substance instead of pharos_drug, right? I think we want to use the standard bioentity types, not our own inventions. I would suggest we fix that now, but if better to wait and global search/replace later, okay.

**Comment by dkoslicki** [comment]
@edeutsch Ok, I'll start on that now. I'd prefer just a uniform labeling of questions (Q0, Q1,...). Not having special cases is best I think. I say since python starts counting at 0, we can too :)

As for disont_disease vs disease, yes, this will be done with a global search/replace (Steve gave me the mapping in #34). If I try to do it now on master, everything will break, so I'm of a mind to get things working on this end just how we like them before updating to the standardized KG.

**Comment by dkoslicki** [comment]
@edeutsch quick question: how will you know which script to call if all RTXQuery gets is known_query_type_ID=="Q34" and a dict of terms?

**Comment by edeutsch** [comment]
great!

**Comment by edeutsch** [comment]
RTXQuery will need to open your Questions.tsv file and get the information out of there. If you can provide a nice class for me that automatically reads the file for me otherwise allows me to pull information out that would be great. Ideally both RTXQuery and the QuestionTranslator would use the same class. I'm thinking I'd like to do this:

import QueryType
queryType = QueryType(known_query_type_ID)
executionString = queryType.executionString
executionString = re.sub(........)

that's it! RTXQuery shouldn't need to know anything else.
It would then do a searchreplace using the incoming terms into the executionString and run it.

Does that seem sensible?

**Comment by dkoslicki** [comment]
Yup, got it!

**Comment by dkoslicki** [comment]
Just a note: only Q0 blindly takes text from the user query. So "What is rm -rf /home" would return ``{"terms":{"term":"rm -rf /home"}}``. All the other questions populate individual parameters only with node names and node/relationship types. So unless a node in the graph is named "rm -rf .", we should (hopefully) be able to trust the terms in the QX result for X>0.

**Comment by edeutsch** [comment]
yes, right. BUT, currently Q0 doesn't use the shell, so it is immune from such shell-based attacks. Maybe this is a good reason NOT to switch Q0 to a shell-based system. And incentive to move the others to a non-shell based system.

regarding second point, I'm less concerned about the QuestionParser than the query service. Using the query service, the user is able to send terms directly. And those terms could contain malicious stuff. I suppose this brings to the fore the point that just because translate/ is vetting the terms, it doesn't mean that query/ should assume that the terms of good and valid.

**Comment by dkoslicki** [comment]
Ok, the changes have been implemented. You can now create execution strings with:
```python
import ParseQuestion
p = ParseQuestion.ParseQuestion()
input_question = #whatever
res = p.format_response(input_question)
execution_string = p.get_execution_string(res["known_query_type_id"], res["terms"])
```
What this function does is map the known_query_type_id to the correct Question class. This question class has a property that is a string Template. This template is then ``.safe_substitute(parameters)`` to insert the terms in the string. There is a bit of error checking included: making sure it's an actual _known_ query type id, and making sure all parameters were populated. Throws an exception otherwise.

**Comment by edeutsch** [comment]
This looks great on the surface. The class name and method name seem a little strange, but that's trivial detail. I will try to test soon, but I am a little too frantic with preparations for tomorrows departure to do any coding now.

Thank you!
